### DISCUSSION TITLE
Loader data generics

### DISCUSSION THREAD

**Discussion by siric** [remix-run/react-router#9854] [Proposals]
### What is the new or updated feature that you are suggesting?

Instead of returning `unknown` by default and then typecasting (as seen in one of the [examples](https://github.com/remix-run/react-router/blob/f0693f5c792ff42fe501c638fc76921a66db8c80/examples/data-router/src/routes.tsx#L95)), prefer generics:

Bad
```js  
const data = useLoaderData() as Data
```

Good
```js  
const data = useLoaderData<Data>()
```

The same should be applied to `useRouteLoaderData` and `useActionData` and it should account for `defer` as well (see #9641 and #9642).

### Why should this feature be included?

Generics over typecasting.

**Comment by devpeerapong** [comment]
[This will be decided in 6.5](https://github.com/remix-run/react-router/pull/9070#discussion_r923470860)

**Comment by brophdawg11** [comment]
I'm going to convert this to a discussion so it can go through our new [Open Development](https://remix.run/blog/open-development) process.  Please upvote the new Proposal if you'd like to see this considered!

**Comment by brophdawg11** [comment]
Related PR to ensure we have typings on `json` and `redirect`: https://github.com/remix-run/react-router/pull/9669

**Reply by brophdawg11** [reply]
Related PR: https://github.com/remix-run/react-router/pull/9670

**Reply by brophdawg11** [reply]
Related Issue and PR for `defer` types as well: https://github.com/remix-run/react-router/issues/9641 https://github.com/remix-run/react-router/pull/9642

**Reply by lennerd** [reply]
@brophdawg11 thanks for linking the PRs. I'm a bit confused though. Thesse PRs are all either closed or merged.

Maybe I'm wrong but there is still no support for generics in `useLoaderData` or `useActionData`, right? What is the state of the discussion here?

**Reply by brophdawg11** [reply]
Yeah those are closed PRs we may refer to if we implement this.  There are no generics at the moment - we may still do them but are also considering some alternative ways to provide stronger type safety/inference.  See https://github.com/remix-run/react-router/discussions/9687 and https://github.com/remix-run/react-router/issues/9768.

**Comment by jamesopstad** [comment]
I've been working on a typesafe wrapper library for React Router (https://github.com/jamesopstad/typesafe-router) and would like to build in better support for the various data types that can be returned by loaders. Currently, `defer`, `redirect` and `json` are not generic so type inference is lost when using these functions (this isn't strictly necessary for `redirect`). If we take `defer` as an example then the `data` parameter would be generic in the `defer` function and `DeferredData` class. Ideally, `defer`, `redirect` and `json` would return branded types so that they can be differentiated and the correct type can be extracted for each.

I'd be happy to put together a PR if you are in favour of this approach. It would also benefit regular React Router users.

**Comment by danielnixon** [comment]
> Bad
> ```ts
> const data = useLoaderData() as Data
> ```
> Good
> ```ts
> const data = useLoaderData<Data>()
> ```

Doesn't this just hide the `as` type assertion in the implementation of `useLoaderData`? There's still the same number of type assertions overall (and the level of type safety hasn't increased or decreased), it's just that we've disguised the type assertion.

As they say in https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

> getMeAT<T>(): T: If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion. Prefer to use a real type assertion, e.g. getMeAT() as number. Example where a type parameter is acceptable: function id<T>(value: T): T;. Example where it is not acceptable: function parseJson<T>(json: string): T;. Exception: new Map<string, number>() is OK.

Overall you might even argue that we're worse off, because disguising the type assertion in this way renders it invisible from tooling such as https://github.com/plantain-00/type-coverage that would otherwise be able to warn us about the `as` in the AST.

**Reply by brophdawg11** [reply]
> Doesn't this just hide the as type assertion

Yes, absolutely.  We argued this _repeatedly_ over in Remix and eventually gave in.  We don't love it though and now that we've got a better grasp on our patterns (and TS in general!) we're focusing on better ways to do this in future APIs.  That's part of why we've avoided carrying this over from Remix to React Router.

**Reply by danielnixon** [reply]
Thanks for holding the line. Using the same approach as [useContext](https://react.dev/reference/react/useContext#usecontext) is perhaps the answer. It's not completely type-unsafe and it comes from React itself, so...

**Reply by pcattori** [reply]
Agreed. Generics with hidden type casts end up turning off type-checking rather than enforcing type-safety. If you want inferred type-safety for loader data you can use [typegen'd `Route`](https://reactrouter.com/explanation/type-safety) and pass through typed props OR you can use the [upcoming `useRoute(routeId?: string)`(https://github.com/remix-run/react-router/discussions/14417).

If you still need to be able to ergonomically cast types, we could provide a `RouteInfo<RouteId>` type utility:

```ts
const data = useLoaderData() as RouteInfo<"routes/home">['loaderData']
```

In this case its a feature, not a bug, that the `as` cast is explicit.

**Comment by JoepKockelkorn** [comment]
@brophdawg11 Do you have any update on this? I've looked at [the Roadmap](https://github.com/orgs/remix-run/projects/5/views/1) but it doesn't seem completely up-to-date and can't find this issue on it.

**Comment by silvenon** [comment]
I was quite surprised to not find `useLoaderData<typeof loader>()` in React Router. Are there plans to port it over from Remix?

**Comment by tykhan** [comment]
Generics should also work with `useAsyncValue()`

**Comment by joernroeder** [comment]
any updates on this? seems like a strange gap in the toolkit to have…

**Comment by pcattori** [comment]
Closing this as we would rather provide a new type-safe `useRoute` hook and potentially a type utility like `RouteInfo<RouteId>` for explicit type casts.

For details see: https://github.com/remix-run/react-router/discussions/9854#discussioncomment-14587081
